src/react/hooks/useReadQuery.ts

```diff
@@ -1,49 +1,43 @@
-import * as React from "react";
-import invariant from "ts-invariant";
-import { NetworkStatus } from "../../core/index.js";
-import { QUERY_REFERENCE_SYMBOL, type QueryReference } from "../cache/QueryReference.js";
-import { __use } from "./internal/index.js";
-import { toApolloError } from "./useSuspenseQuery.js";
+import * as React from 'react';
+import {
+  unwrapQueryRef,
+  type QueryReference,
+} from '../cache/QueryReference.js';
+import { __use } from './internal/index.js';
+import { toApolloError } from './useSuspenseQuery.js';
+import { invariant } from '../../utilities/globals/index.js';
+import { useSyncExternalStore } from './useSyncExternalStore.js';
 
 export function useReadQuery<TData>(queryRef: QueryReference<TData>) {
-  const [, forceUpdate] = React.useState(0);
-  const internalQueryRef = queryRef[QUERY_REFERENCE_SYMBOL];
+  const internalQueryRef = unwrapQueryRef(queryRef);
   invariant(
     internalQueryRef.promiseCache,
     'It appears that `useReadQuery` was used outside of `useBackgroundQuery`. ' +
       '`useReadQuery` is only supported for use with `useBackgroundQuery`. ' +
       'Please ensure you are passing the `queryRef` returned from `useBackgroundQuery`.'
   );
 
-  const skipResult = React.useMemo(() => {
-    const error = toApolloError(internalQueryRef.result);
+  const { promiseCache, key } = internalQueryRef;
 
-    return {
-      loading: false,
-      data: internalQueryRef.result.data,
-      networkStatus: error ? NetworkStatus.error : NetworkStatus.ready,
-      error,
-    };
-  }, [internalQueryRef.result]);
-
-  let promise = internalQueryRef.promiseCache.get(internalQueryRef.key);
-
-  if (!promise) {
-    promise = internalQueryRef.promise;
-    internalQueryRef.promiseCache.set(internalQueryRef.key, promise);
+  if (!promiseCache.has(key)) {
+    internalQueryRef.promiseCache.set(key, internalQueryRef.promise);
```

Comment:
```suggestion
  if (!promiseCache.has(key)) {
    promiseCache.set(key, internalQueryRef.promise);
```

Is there a real need for this logic?

I think it could make more sense to have a `readQuery` function:

```suggestion
const readQuery = () => promiseCache.get(key) || internalQueryRef.promise
```
and use that instead of `() => promiseCache.get(key)!` down in the `uSES` call.

Reply:
I'm a little leery to make this change until we get a few more tests written for `useBackgroundQuery`/`useReadQuery`. Setting the value here is meant to be reflected back to `useBackgroundQuery` where `promiseCache` state actually lives. When calling `refetch` or `fetchMore`, the `promiseCache` is copied to allow React to fork the state for `startTransition` and `useDeferredValue`. This particular `promiseCache.set` is handle when `variables` change and allows us to write the value to the promise cache without the need for an additional `set`. Hopefully that makes sense.

This is tracked in #10879

Reply:
Okay, then I was just missing a lot of context here :)

What about this suggestion though?

```suggestion
  if (!promiseCache.has(key)) {
   promiseCache.set(key, internalQueryRef.promise);
```

Reply:
Oops, yes. No idea why I didn't use the destructured value 😅

Reply:
[99c2881](https://github.com/apollographql/apollo-client/pull/11112/commits/99c288119d44c66f9b874d5307c5e6607708a070)